opm/autodiff/BlackoilAquiferModel_impl.hpp

```diff
@@ -1,111 +1,141 @@
+        }
+    }
+
+    template<typename TypeTag>
+    void
+    BlackoilAquiferModel<TypeTag>::beginIteration()
+    { }
+
+    template<typename TypeTag>
+    template <class Context>
+    void
+    BlackoilAquiferModel<TypeTag>::addToSource(RateVector& rates,
+                                               const Context& context,
+                                               unsigned spaceIdx,
+                                               unsigned timeIdx) const
+    {
+        for (auto& aquifer: aquifers_) {
+            aquifer.addToSource(rates, context, spaceIdx, timeIdx);
+        }
+    }
+
+    template<typename TypeTag>
+    void
+    BlackoilAquiferModel<TypeTag>::endIteration()
+    { }
+
+    template<typename TypeTag>
+    void
+    BlackoilAquiferModel<TypeTag>::endTimeStep()
+    {
+        for (auto aquifer = aquifers_.begin(); aquifer != aquifers_.end(); ++aquifer) {
+            aquifer->endTimeStep();
+        }
+    }
+
+    template<typename TypeTag>
+    void
+    BlackoilAquiferModel<TypeTag>::endEpisode()
+    { }
+
+    // Initialize the aquifers in the deck
+    template<typename TypeTag>
+    void
+    BlackoilAquiferModel<TypeTag>:: init()
+    {
+        const auto& deck = this->simulator_.vanguard().deck();
+
+        if ( !deck.hasKeyword("AQUCT") ) {
+            return ;
+        }
+
+        //updateConnectionIntensiveQuantities();
+        const auto& eclState = this->simulator_.vanguard().eclState();
+
+        // Get all the carter tracy aquifer properties data and put it in aquifers vector
+        const AquiferCT aquiferct = AquiferCT(eclState,deck);
+        const Aquancon aquifer_connect = Aquancon(eclState.getInputGrid(), deck);
+
+        std::vector<AquiferCT::AQUCT_data> aquifersData = aquiferct.getAquifers();
+        std::vector<Aquancon::AquanconOutput> aquifer_connection = aquifer_connect.getAquOutput();
+
+        assert( aquifersData.size() == aquifer_connection.size() );
+
+
+        // generate the mapping from Cartesian coordinates to the index in the actual grid
+        const auto& ugrid = simulator_.vanguard().grid();
+        const auto& cart_dims = Opm::UgGridHelpers::cartDims(ugrid);
+        setupCartesianToCompressed(Opm::UgGridHelpers::globalCell(ugrid),
+                                   cart_dims[0] * cart_dims[1] * cart_dims[2]);
+
+        for (size_t i = 0; i < aquifersData.size(); ++i)
+        {
+            aquifers_.push_back(
+               AquiferCarterTracy<TypeTag> (aquifersData.at(i),
+                    aquifer_connection.at(i), cartesian_to_compressed_, this->simulator_)
+                               );
+        }
+
+
+    }
+
+    template<typename TypeTag>
+    bool
+    BlackoilAquiferModel<TypeTag>:: aquiferActive() const
+    {
+        return !aquifers_.empty();
+    }
+
+
+
+    template<typename TypeTag>
+    void
+    BlackoilAquiferModel<TypeTag>::
+    setupCartesianToCompressed(const int* global_cell, const unsigned number_of_cartesian_cells)
```

Comment:
This is the function, which is copied from the `BlackoilWellModel_impl.hpp`.

Reply:
Please try to avoid this.

We already have many too many implementations of this operation.  I'd much prefer if this were centralised to a common location&mdash;for instance as a companion component to OPM-grid's [`compressedToCartesian()`](https://github.com/OPM/opm-grid/blob/2ec5099c0dace80dc924702e66afe3569096a914/opm/grid/utility/compressedToCartesian.hpp) function.

Reply:
There are  two different implementation for the functionality  `cartesian_to_compressed` in OPM now. 

One is to treat `cartesian_to_compressed` as a map due to its possibly sparse structure, the other way is to treat it as a `vector`, with `-1` fills the inactive grid blocks. With the second approach, we need to know the size of the full grid. 

Which approach is suggested if we implement `a companion component to OPM-grid's compressedToCartesian() function.`

I am okay with  either way.  Please suggest.

Reply:
Regarding `std::vector<>` vs. `std::map<>`.  The former is constant time lookup, the latter usually requires less memory.  If we're thinking about a general facility, I might actually consider a  `std::unordered_map<int,int>` with a load factor less than 0.9 to balance lookup-speed and memory requirements.  I'd prefer to avoid `std::map<>` due to the linked list nature of the lookup.

I should also mention [`EclipseGrid::activeIndex()`](https://github.com/OPM/opm-common/blob/cf61068ad1b748793dd3ae340ed99df7cf93111f/opm/parser/eclipse/EclipseState/Grid/EclipseGrid.hpp#L88-L89) which might be relevant in this area.

Reply:
`EclipseGrid::activeIndex() ` might not consider parallel? I can be wrong.